cdap-app-templates/cdap-program-report/cdap-program-report-base/src/main/java/co/cask/cdap/report/ReportGenerationSpark.java

```diff
@@ -176,43 +234,103 @@ private ReportList getReportList(int offset, int limit) throws Exception {
       return new ReportList(offset, limit, reportIdDirs.size(), reportStatuses);
     }
 
+    @POST
+    @Path("/reports/{report-id}/share")
+    public void shareReport(HttpServiceRequest request, HttpServiceResponder responder,
+                            @PathParam("report-id") String reportId) {
+      Location reportIdDir;
+      try {
+        String userName = getUserName(request.getAllHeaders());
+        reportIdDir = getDatasetBaseLocation(ReportGenerationApp.REPORT_FILESET).append(userName).append(reportId);
+        if (!reportIdDir.exists()) {
+          responder.sendError(404, String.format("Invalid report-id %s, report does not exist", reportId));
+          return;
+        }
+        String shareId = encodeShareId(new ReportIdentifier(userName, reportId));
+        responder.sendJson(200, new ShareId(shareId), ShareId.class, GSON);
+      } catch (IOException | GeneralSecurityException e) {
+        LOG.error("Failed to read report with id {}", reportId, e);
+        responder.sendError(500, String.format("Failed to read report with id %s because of error: %s",
+                                               reportId, e.getMessage()));
+        return;
+      }
+    }
+
+    private String encodeShareId(ReportIdentifier reportIdentifier) throws GeneralSecurityException, IOException {
+      byte[] response = GSON.toJson(reportIdentifier).getBytes(StandardCharsets.UTF_8);
+      if (encryptionCipher == null) {
+        encryptionCipher = initializeCipher(Cipher.ENCRYPT_MODE);
+      }
+      byte[] cipherReponse = encryptionCipher.doFinal(response);
```

Comment:
Looks like `Cipher` is not thread safe, so the instances cannot be shared among multiple threads.

Reply:
Its not shared among multiple threads here too, as multiple handler instances will run on multiple threads and they will have their own Cipher instance. let me know if my understanding is incorrect.. It might be better to create this cipher variable and instance inside this method to avoid the confusion.